[svg] setting `d: none` on a `<path>` does not override the `d` SVG attribute

graouts opened the pull request:
#### e33f42cd046adc916beddbf75157f692f6f22ad7
<pre>
[svg] setting `d: none` on a `&lt;path&gt;` does not override the `d` SVG attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=276653">https://bugs.webkit.org/show_bug.cgi?id=276653</a>

Reviewed by Nikolas Zimmermann.

If the `d` CSS property is set to `none`, the resulting `SVGRenderStyle::d()` value is `nullptr`.
We must account for this in `SVGPathElement::path()` and `SVGPathElement::pathByteStream()` and
return an empty path in the case that we don&apos;t have a `BasicShapePath` set on the style. In fact,
at long as the `d` CSS property is enabled, we should never return the SVG attribute value.

Making this code change alone yielded two test regressions:

    - svg/custom/svg-curve-with-relative-coordinates.html
    - svg/animations/path-animation.svg

Indeed, neither of those two tests explicitly sets the `d` SVG attribute, relying instead on `&lt;animate&gt;` or
the `pathSegList` DOM API to set the path data. The lack of a `d` SVG attribute means the traditional mechanism
to compile a list of presentation attributes for a `&lt;path&gt;` attribute will not pick up the `d` property.

So we add a new method to collect additional presentational hints as was already done for the HTML `&lt;img&gt;`
element, promoting the `collectExtraStyleForPresentationalHints()` method from `HTMLImageElement` to a virtual
method on `StyledElement` and making `SVGPathElement` override it. Using this method, we can collect the `d`
presentational hint if it hasn&apos;t already been picked up.

Finally, since the lack of accounting for `d: none` when rendering wasn&apos;t caught by existing tests, we add
a new WPT reftest.

* LayoutTests/imported/w3c/web-platform-tests/svg/animations/spaces-at-end-of-path-data-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/path/property/d-none-expected.svg: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/path/property/d-none-ref.svg: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/path/property/d-none.svg: Added.
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::rebuildPresentationalHintStyle):
* Source/WebCore/dom/StyledElement.h:
(WebCore::StyledElement::collectExtraStyleForPresentationalHints):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/svg/SVGPathByteStream.h:
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::pathByteStream const):
(WebCore::SVGPathElement::path const):
(WebCore::SVGPathElement::collectPresentationalHintsForAttribute):
(WebCore::SVGPathElement::collectExtraStyleForPresentationalHints):
(WebCore::SVGPathElement::collectDPresentationalHint):
* Source/WebCore/svg/SVGPathElement.h:

Canonical link: <a href="https://commits.webkit.org/281041@main">https://commits.webkit.org/281041@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c11eb368cffd1f24d7529213689d6808e7cb12d8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58330 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37658 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10814 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61955 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8775 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60459 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45294 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8970 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47255 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6267 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60361 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35292 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50436 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28096 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32051 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7734 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7779 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54002 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8005 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63660 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2244 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8049 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54572 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2252 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50450 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54637 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1914 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8724 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33487 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34573 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35657 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34318 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann reviewed (approved):
Nice fix 👍 If you are fine with 'virtualizing' the collectExtra...() method (potential overhead for non-HTML <img> / SVG <path> usage)  then go for it 😀 r=me

graouts committed:
[svg] setting `d: none` on a `<path>` does not override the `d` SVG attribute
https://bugs.webkit.org/show_bug.cgi?id=276653

Reviewed by Nikolas Zimmermann.

If the `d` CSS property is set to `none`, the resulting `SVGRenderStyle::d()` value is `nullptr`.
We must account for this in `SVGPathElement::path()` and `SVGPathElement::pathByteStream()` and
return an empty path in the case that we don't have a `BasicShapePath` set on the style. In fact,
at long as the `d` CSS property is enabled, we should never return the SVG attribute value.

Making this code change alone yielded two test regressions:

    - svg/custom/svg-curve-with-relative-coordinates.html
    - svg/animations/path-animation.svg

Indeed, neither of those two tests explicitly sets the `d` SVG attribute, relying instead on `<animate>` or
the `pathSegList` DOM API to set the path data. The lack of a `d` SVG attribute means the traditional mechanism
to compile a list of presentation attributes for a `<path>` attribute will not pick up the `d` property.

So we add a new method to collect additional presentational hints as was already done for the HTML `<img>`
element, promoting the `collectExtraStyleForPresentationalHints()` method from `HTMLImageElement` to a virtual
method on `StyledElement` and making `SVGPathElement` override it. Using this method, we can collect the `d`
presentational hint if it hasn't already been picked up.

Finally, since the lack of accounting for `d: none` when rendering wasn't caught by existing tests, we add
a new WPT reftest.

* LayoutTests/imported/w3c/web-platform-tests/svg/animations/spaces-at-end-of-path-data-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/svg/path/property/d-none-expected.svg: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/path/property/d-none-ref.svg: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/path/property/d-none.svg: Added.
* Source/WebCore/dom/StyledElement.cpp:
(WebCore::StyledElement::rebuildPresentationalHintStyle):
* Source/WebCore/dom/StyledElement.h:
(WebCore::StyledElement::collectExtraStyleForPresentationalHints):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/svg/SVGPathByteStream.h:
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::pathByteStream const):
(WebCore::SVGPathElement::path const):
(WebCore::SVGPathElement::collectPresentationalHintsForAttribute):
(WebCore::SVGPathElement::collectExtraStyleForPresentationalHints):
(WebCore::SVGPathElement::collectDPresentationalHint):
* Source/WebCore/svg/SVGPathElement.h:

Canonical link: https://commits.webkit.org/281041@main
